### ISSUE TITLE
boxes.c: 9 * resource leak ?

### ISSUE THREAD

**Issue by dcb314** [ascii-boxes/boxes#28]
[boxes.c:422]: (error) Resource leak: new_yyin
[boxes.c:426]: (error) Resource leak: new_yyin
[boxes.c:431]: (error) Resource leak: new_yyin
[boxes.c:451]: (error) Resource leak: new_yyin
[boxes.c:457]: (error) Resource leak: new_yyin
[boxes.c:511]: (error) Resource leak: new_yyin
[boxes.c:515]: (error) Resource leak: new_yyin
[boxes.c:521]: (error) Resource leak: new_yyin
[boxes.c:704]: (error) Resource leak: f

**Comment by tsjensen** [comment]
I'm sorry, but we would need some more context. Are you reporting a bug? If so, how can it be reproduced? What is your environment?

**Comment by dcb314** [comment]
>Are you reporting a bug?

Yes.

>If so, how can it be reproduced?

Use C/C++ static analyser cppcheck over the source code of the boxes product.

>What is your environment?

Linux.

**Comment by tsjensen** [comment]
I understand that you ran a static analysis tool on the boxes source and it output something. I'm sorry, but it is really hard to act on a general statement like that. The given [line 422](https://github.com/ascii-boxes/boxes/blob/master/src/boxes.c#L422) contains a switch statement. I don't see how this is a bug. What is the problem? Are you getting an error message?

**Comment by dcb314** [comment]
1.

[boxes.c:422]: (error) Resource leak: new_yyin

Source code is

        new_yyin = fopen (s, "r");
        if (new_yyin == NULL) {
            fprintf (stderr, "%s: Couldn't open config file '%s' "
                    "for input (taken from $BOXES).\n", PROJECT, s);
            return 1;
        }
        rc = is_dir (s);
        if (rc == -1)
            return 1;

Hopefully the leak is clear from the code, when rc == -1.

2.

[boxes.c:426]: (error) Resource leak: new_yyin

Source code is

        else if (rc) {
            fprintf (stderr, "%s: Alleged config file '%s' is a directory "
                    "(taken from $BOXES)\n", PROJECT, s);
            return 1;
        }

and so on downto line 521.

        if (yyfilename == NULL) {
            perror (PROJECT);
            return 1;
        }

For line 704, I have the following code

              f = fopen (optarg, "r");
                if (f == NULL) {
                    fprintf (stderr, "%s: Couldn\'t open config file \'%s\' "
                            "for input.\n", PROJECT, optarg);
                    return 1;
                }
                rc = is_dir (optarg);
                if (rc == -1)
                    return 1;

Where hopefully the leak is again clear.

In general, every fopen should have a matching fclose.

$ egrep "fopen|fclose" ../BUILD/boxes-1.1.1/src/boxes.c
 *  On Windows, this check seems unnecessary, because fopen() seems to fail
        new_yyin = fopen (s, "r");
        new_yyin = fopen (BOXES_CONFIG, "r");
                    fclose (new_yyin);
        new_yyin = fopen (exepath, "r");
        new_yyin = fopen (GLOBALCONF, "r");
                f = fopen (optarg, "r");
            opt.infile = fopen (argv[optind], "r");
            opt.outfile = fopen (argv[optind+1], "w");
                    fclose (opt.infile);
$

**Comment by tsjensen** [comment]
Thanks for the description. I see that there are some `fclose()` calls missing. It's not really a *leak*, because the entire process normally terminates less than a second later, closing all handles. But I see that it would be cleaner to close handles more diligently.

Can you do a PR for that?

**Comment by livibetter** [comment]
@tsjensen I am thinking to submit a PR for this issue, do you mind if I use like `goto err1;` for most of them? If adding `fclose` to where errors are, brackets are required. Which one do you prefer?

**Comment by tsjensen** [comment]
Thank you for your offer! I'd much appreciate a PR. Even though I feel no pain because of this (as explained above), it would improve the code base.
I don't understand you questions, I'm afraid. (`goto` is generally considered harmful. Curly braces (`{}`) are fine.)

**Comment by livibetter** [comment]
@tsjensen I have same thought about those leaks, but I hate to see an issue left open, especially when it's just trivial task, that's why I am doing this.

And I know some people would try to avoid `goto`, that's why I wanted to ask first. That propose was going like this:

```c
    new_yyin = fopen (s, "r");
    // ...
    rc = is_dir (s);
    if (rc == -1)
        goto err1;  // <- return 1;
// ...
err1:
    fclose(new_yyin);
    return 1;
```

There are several which could be done like this, but I will go for braces in the PR.

**Comment by tsjensen** [comment]
Fixed by #37
